src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -0,0 +1,403 @@
+                            //       So look for deletes first since those will mean the listening app can't do anything with the item since it doesn't exist (such as stat 
+                            //       the file to determine what exactly changed, in this example).
+                            if ((eventFlags[i] & Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemRemoved) == Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemRemoved)
+                            {
+                                NotifyFileSystemEventArgs(WatcherChangeTypes.Deleted, relativePath);
+                            }
+                            else if ((eventFlags[i] & Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemCreated) == Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemCreated)
+                            {
+                                // Next look for creates since a create + modification coalesced could confuse apps since a file they haven't heard of yet would get a mod event
+                                NotifyFileSystemEventArgs(WatcherChangeTypes.Created, relativePath);
+                            }
+                            else if (((eventFlags[i] & Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemInodeMetaMod) == Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemInodeMetaMod) ||
+                                     ((eventFlags[i] & Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemModified) == Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemModified) ||
+                                     ((eventFlags[i] & Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemFinderInfoMod) == Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemChangeOwner) ||
+                                     ((eventFlags[i] & Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemXattrMod) == Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemXattrMod))
+                            {
+                                // Everything else is a modification
+                                NotifyFileSystemEventArgs(WatcherChangeTypes.Changed, relativePath);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        private bool ShouldRescanOccur(Interop.CoreServices.FSEventStreamEventFlags flags)
+        {
+            // Check if the MustScanSubDirs flag exists
+            bool shouldRescan = ((flags & Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagMustScanSubDirs) == Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagMustScanSubDirs);
```

Comment:
Not trying to derail this comment, but you can probably add code highlighting in those blocks to make them a tad easier to read:
https://support.codebasehq.com/articles/tips-tricks/syntax-highlighting-in-markdown